File: scripts/audit-web-aria.ts
#!/usr/bin/env tsx
/**
 * Web ARIA Labels Audit Script
 *
 * Scans all web components for ARIA label compliance using regex patterns.
 * Checks for:
 * - aria-label or aria-labelledby on interactive elements
 * - Form inputs with associated labels
 * - Modal/dialog ARIA attributes
 * - Icon buttons with descriptive labels
 */

import { readFileSync, readdirSync, statSync, writeFileSync, mkdirSync } from 'fs';
import { join, relative } from 'path';

interface AriaViolation {
  file: string;
  line: number;
  element: string;
  issue: string;
  severity: 'error' | 'warning';
  recommendation: string;
  codeSnippet: string;
}

interface AuditResult {
  totalComponents: number;
  violations: AriaViolation[];
  warnings: AriaViolation[];
  byComponent: Record<string, {
    violations: AriaViolation[];
    warnings: AriaViolation[];
  }>;
}

const COMPONENTS_DIR = join(process.cwd(), 'apps/web/src/components');
const EXCLUDED_DIRS = ['__tests__', 'node_modules', '.next', 'dist', 'build'];
const EXCLUDED_FILES = ['.test.tsx', '.spec.tsx', '.stories.tsx'];

function getAllComponentFiles(dir: string, fileList: string[] = []): string[] {
  try {
    const files = readdirSync(dir);

    for (const file of files) {
      const filePath = join(dir, file);
      const stat = statSync(filePath);

      if (stat.isDirectory()) {
        if (!EXCLUDED_DIRS.includes(file)) {
          getAllComponentFiles(filePath, fileList);
        }
      } else if (file.endsWith('.tsx') && !EXCLUDED_FILES.some(excluded => file.includes(excluded))) {
        fileList.push(filePath);
      }
    }
  } catch {
    // Directory might not exist or be inaccessible
  }

  return fileList;
}

function auditFile(filePath: string): AriaViolation[] {
  const violations: AriaViolation[] = [];
  const relativePath = relative(process.cwd(), filePath);
  let content: string;

  try {
    content = readFileSync(filePath, 'utf-8');
  } catch {
    return violations;
  }

  const lines = content.split('\n');

  // Pattern 1: Icon buttons without aria-label
  // Match: <Button size="icon" or <IconButton or buttons with icon classes
  const iconButtonPattern = /<(Button|IconButton|button)[^>]*(size\s*=\s*["']icon["']|className[^>]*icon)[^>]*>/gi;
  let match;
  while ((match = iconButtonPattern.exec(content)) !== null) {
    const lineNumber = content.substring(0, match.index).split('\n').length;
    const elementContent = match[0];

    // Check if aria-label exists
    if (!/aria-label\s*=/i.test(elementContent)) {
      // Check if it's followed by aria-label in next lines (multiline)
      const nextLines = lines.slice(lineNumber - 1, Math.min(lineNumber + 5, lines.length));
      const hasAriaLabelNearby = nextLines.some(line => /aria-label\s*=/i.test(line));

      if (!hasAriaLabelNearby) {
        violations.push({
          file: relativePath,
          line: lineNumber,
          element: 'IconButton/Button',
          issue: 'Icon button missing aria-label',
          severity: 'error',
          recommendation: 'Add aria-label prop with descriptive text (e.g., "Close dialog", "Send message")',
          codeSnippet: elementContent.substring(0, 100),
        });
      }
    }
  }

  // Pattern 2: Input elements without labels
  const inputPattern = /<(input|textarea|select)([^>]*?)>/gi;
  while ((match = inputPattern.exec(content)) !== null) {
    const lineNumber = content.substring(0, match.index).split('\n').length;
    const elementContent = match[0];
    const attributes = match[2] || '';

    // Skip if it's a hidden input
    if (/type\s*=\s*["']hidden["']/i.test(attributes)) {
      continue;
    }

    // Check for aria-label, aria-labelledby, or id (which might be associated with label)
    const hasAriaLabel = /aria-label\s*=/i.test(attributes);
    const hasAriaLabelledBy = /aria-labelledby\s*=/i.test(attributes);
    const _hasId = /id\s*=\s*["'][^"']+["']/i.test(attributes);

    // Check if there's a label element with htmlFor pointing to this input
    const idMatch = attributes.match(/id\s*=\s*["']([^"']+)["']/i);
    const hasLabel = idMatch ? new RegExp(`htmlFor\\s*=\\s*["']${idMatch[1]}["']`, 'i').test(content) : false;

    if (!hasAriaLabel && !hasAriaLabelledBy && !hasLabel) {
      violations.push({
        file: relativePath,
        line: lineNumber,
        element: match[1],
        issue: 'Form input missing label or aria-label',
        severity: 'error',
        recommendation: 'Add htmlFor/id relationship, aria-label, or aria-labelledby',
        codeSnippet: elementContent.substring(0, 100),
      });
    }
  }

  // Pattern 3: Buttons without aria-label or visible text
  const buttonPattern = /<(Button|button)([^>]*?)>(.*?)(?:<\/Button>|<\/button>)/gis;
  while ((match = buttonPattern.exec(content)) !== null) {
    const lineNumber = content.substring(0, match.index).split('\n').length;
    const elementContent = match[0];
    const attributes = match[2] || '';
    const children = match[3] || '';

    // Skip if it has aria-label
    if (/aria-label\s*=/i.test(attributes)) {
      continue;
    }

    // Skip if it's an icon button (handled above)
    if (/size\s*=\s*["']icon["']/i.test(attributes)) {
      continue;
    }

    // Check if button has visible text content
    const hasText = /[A-Za-z0-9]{2,}/.test(children.replace(/<[^>]+>/g, '').trim());

    // Check for icon-only buttons (common patterns)
    const isIconOnly = /<[A-Z][^>]*Icon[^>]*>|<svg/i.test(children) && !hasText;

    if (isIconOnly || (!hasText && children.trim().length < 10)) {
      violations.push({
        file: relativePath,
        line: lineNumber,
        element: 'Button',
        issue: 'Button missing aria-label or visible text',
        severity: 'warning',
        recommendation: 'Add aria-label prop or ensure button has visible text content',
        codeSnippet: elementContent.substring(0, 150),
      });
    }
  }

  // Pattern 4: Modal/Dialog elements without proper ARIA
  const modalPattern = /<(Dialog|DialogContent|Modal|dialog)([^>]*?)>/gi;
  while ((match = modalPattern.exec(content)) !== null) {
    const lineNumber = content.substring(0, match.index).split('\n').length;
    const elementContent = match[0];
    const attributes = match[2] || '';

    const issues: string[] = [];
    if (!/role\s*=\s*["']dialog["']/i.test(attributes) && !/role\s*=\s*["']alertdialog["']/i.test(attributes)) {
      issues.push('Missing role="dialog"');
    }
    if (!/aria-modal\s*=\s*["']true["']/i.test(attributes)) {
      issues.push('Missing aria-modal="true"');
    }
    if (!/aria-labelledby\s*=/i.test(attributes)) {
      issues.push('Missing aria-labelledby');
    }

    if (issues.length > 0) {
      violations.push({
        file: relativePath,
        line: lineNumber,
        element: match[1],
        issue: `Modal/dialog missing ARIA attributes: ${issues.join(', ')}`,
        severity: 'error',
        recommendation: 'Add role="dialog", aria-modal="true", aria-labelledby, and optionally aria-describedby',
        codeSnippet: elementContent.substring(0, 150),
      });
    }
  }

  // Pattern 5: Links without accessible text
  const linkPattern = /<(a|Link)([^>]*?)>(.*?)(?:<\/a>|<\/Link>)/gis;
  while ((match = linkPattern.exec(content)) !== null) {
    const lineNumber = content.substring(0, match.index).split('\n').length;
    const attributes = match[2] || '';
    const children = match[3] || '';

    // Skip if it has aria-label
    if (/aria-label\s*=/i.test(attributes)) {
      continue;
    }

    // Check if link has visible text or image with alt text
    const hasText = /[A-Za-z0-9]{2,}/.test(children.replace(/<[^>]+>/g, '').trim());
    const hasImageWithAlt = /<img[^>]*alt\s*=\s*["'][^"']+["']/i.test(children);

    if (!hasText && !hasImageWithAlt) {
      violations.push({
        file: relativePath,
        line: lineNumber,
        element: 'Link',
        issue: 'Link missing aria-label or accessible text',
        severity: 'warning',
        recommendation: 'Add aria-label prop or ensure link has visible text or image with alt text',
        codeSnippet: match[0].substring(0, 150),
      });
    }
  }

  // Pattern 6: Interactive elements with onClick but no ARIA
  const onClickPattern = /<([A-Z][a-zA-Z]*|\w+)([^>]*onClick[^>]*?)>/g;
  while ((match = onClickPattern.exec(content)) !== null) {
    const lineNumber = content.substring(0, match.index).split('\n').length;
    const elementName = match[1];
    const attributes = match[2] || '';

    // Skip known interactive elements (button, a, input)
    if (['button', 'a', 'input', 'Button', 'Link'].includes(elementName)) {
      continue;
    }

    // Skip if it has role="button" or aria-label
    if (/role\s*=\s*["']button["']/i.test(attributes) || /aria-label\s*=/i.test(attributes)) {
      continue;
    }

    // Check if it's a known component that handles accessibility
    if (['div', 'span'].includes(elementName.toLowerCase())) {
      violations.push({
        file: relativePath,
        line: lineNumber,
        element: elementName,
        issue: 'Interactive element (onClick) missing role="button" or aria-label',
        severity: 'warning',
        recommendation: 'Add role="button" and aria-label, or use a proper button element',
        codeSnippet: match[0].substring(0, 150),
      });
    }
  }

  return violations;
}

function generateReport(results: AuditResult): string {
  const { totalComponents, violations, warnings, byComponent } = results;
  const errorCount = violations.length;
  const warningCount = warnings.length;

  let report = '# Web ARIA Labels Audit Report\n\n';
  report += `**Generated**: ${new Date().toISOString()}\n\n`;
  report += `## Summary\n\n`;
  report += `- **Total Components Audited**: ${totalComponents}\n`;
  report += `- **Total Violations**: ${errorCount} errors, ${warningCount} warnings\n`;
  report += `- **Compliance Rate**: ${totalComponents > 0 ? ((totalComponents - errorCount) / totalComponents * 100).toFixed(1) : 0}%\n\n`;

  if (errorCount === 0 && warningCount === 0) {
    report += `## ✅ All Clear\n\n`;
    report += `All components have proper ARIA labels and accessibility attributes.\n\n`;
    return report;
  }

  report += `## Violations by Severity\n\n`;

  if (errorCount > 0) {
    report += `### Errors (${errorCount})\n\n`;
    report += `| File | Line | Element | Issue | Recommendation |\n`;
    report += `|------|------|---------|-------|----------------|\n`;

    for (const violation of violations) {
      const fileDisplay = violation.file.replace(/^apps\/web\/src\/components\//, '');
      report += `| \`${fileDisplay}\` | ${violation.line} | \`${violation.element}\` | ${violation.issue} | ${violation.recommendation} |\n`;
    }
    report += `\n`;
  }

  if (warningCount > 0) {
    report += `### Warnings (${warningCount})\n\n`;
    report += `| File | Line | Element | Issue | Recommendation |\n`;
    report += `|------|------|---------|-------|----------------|\n`;

    for (const warning of warnings) {
      const fileDisplay = warning.file.replace(/^apps\/web\/src\/components\//, '');
      report += `| \`${fileDisplay}\` | ${warning.line} | \`${warning.element}\` | ${warning.issue} | ${warning.recommendation} |\n`;
    }
    report += `\n`;
  }

  report += `## Violations by Component\n\n`;

  const componentEntries = Object.entries(byComponent).sort((a, b) => {
    const aTotal = a[1].violations.length + a[1].warnings.length;
    const bTotal = b[1].violations.length + b[1].warnings.length;
    return bTotal - aTotal;
  });

  for (const [component, issues] of componentEntries) {
    if (issues.violations.length === 0 && issues.warnings.length === 0) {
      continue;
    }

    report += `### ${component}\n\n`;
    report += `- **Errors**: ${issues.violations.length}\n`;
    report += `- **Warnings**: ${issues.warnings.length}\n\n`;

    if (issues.violations.length > 0) {
      report += `#### Errors:\n\n`;
      for (const violation of issues.violations) {
        report += `- **Line ${violation.line}**: ${violation.issue}\n`;
        report += `  - Recommendation: ${violation.recommendation}\n`;
        report += `  - Code: \`${violation.codeSnippet}\`\n\n`;
      }
    }

    if (issues.warnings.length > 0) {
      report += `#### Warnings:\n\n`;
      for (const warning of issues.warnings) {
        report += `- **Line ${warning.line}**: ${warning.issue}\n`;
        report += `  - Recommendation: ${warning.recommendation}\n`;
        report += `  - Code: \`${warning.codeSnippet}\`\n\n`;
      }
    }
  }

  report += `## Recommendations\n\n`;
  report += `1. **Icon Buttons**: All icon-only buttons must have \`aria-label\` prop\n`;
  report += `2. **Form Inputs**: All inputs must have associated labels via \`htmlFor\`/\`id\` or \`aria-label\`\n`;
  report += `3. **Modals/Dialogs**: Must have \`role="dialog"\`, \`aria-modal="true"\`, and \`aria-labelledby\`\n`;
  report += `4. **Interactive Elements**: Buttons, links, and custom interactive elements need \`aria-label\` or visible text\n`;
  report += `5. **Live Regions**: Dynamic content should use \`aria-live\` regions for screen reader announcements\n\n`;

  return report;
}

function main(): void {
  console.log('🔍 Starting Web ARIA Labels Audit...\n');

  const files = getAllComponentFiles(COMPONENTS_DIR);
  console.log(`Found ${files.length} component files to audit\n`);

  const results: AuditResult = {
    totalComponents: files.length,
    violations: [],
    warnings: [],
    byComponent: {},
  };

  for (const file of files) {
    const relativePath = relative(process.cwd(), file);
    const componentName = relativePath.split('/').pop()?.replace('.tsx', '') || 'Unknown';

    const violations = auditFile(file);

    const errors = violations.filter(v => v.severity === 'error');
    const warnings = violations.filter(v => v.severity === 'warning');

    results.violations.push(...errors);
    results.warnings.push(...warnings);

    if (errors.length > 0 || warnings.length > 0) {
      results.byComponent[componentName] = {
        violations: errors,
        warnings: warnings,
      };
    }
  }

  console.log(`\n✅ Audit complete!\n`);
  console.log(`- Total violations: ${results.violations.length} errors, ${results.warnings.length} warnings\n`);

  // Generate report
  const report = generateReport(results);
  const reportDir = join(process.cwd(), 'docs/accessibility');
  const reportPath = join(reportDir, 'web-aria-audit-report.md');

  // Ensure docs/accessibility directory exists
  try {
    mkdirSync(reportDir, { recursive: true });
  } catch {
    // Directory might already exist
  }

  writeFileSync(reportPath, report, 'utf-8');
  console.log(`📄 Report saved to: ${reportPath}\n`);

  // Print summary
  if (results.violations.length > 0) {
    console.log(`❌ Found ${results.violations.length} errors that need to be fixed.\n`);
  }
  if (results.warnings.length > 0) {
    console.log(`⚠️  Found ${results.warnings.length} warnings to review.\n`);
  }
  if (results.violations.length === 0 && results.warnings.length === 0) {
    console.log(`✅ No violations found! All components are compliant.\n`);
  }
}

// Run main function
main();

File: scripts/audit-web-aria.py
#!/usr/bin/env python3
"""
Web ARIA Labels Audit Script

Scans all web components for ARIA label compliance using regex patterns.
"""

import os
import re
import json
from pathlib import Path
from typing import List, Dict, Any
from dataclasses import dataclass, asdict

@dataclass
class Violation:
    file: str
    line: int
    element: str
    issue: str
    severity: str
    recommendation: str
    code_snippet: str

def find_component_files(root_dir: str) -> List[str]:
    """Find all component TSX files."""
    files = []
    excluded_dirs = {'__tests__', 'node_modules', '.next', 'dist', 'build'}
    excluded_suffixes = {'.test.tsx', '.spec.tsx', '.stories.tsx'}

    for root, dirs, filenames in os.walk(root_dir):
        # Remove excluded directories
        dirs[:] = [d for d in dirs if d not in excluded_dirs]

        for filename in filenames:
            if filename.endswith('.tsx') and not any(filename.endswith(suffix) for suffix in excluded_suffixes):
                files.append(os.path.join(root, filename))

    return files

def audit_file(file_path: str, root_dir: str) -> List[Violation]:
    """Audit a single file for ARIA violations."""
    violations = []
    relative_path = os.path.relpath(file_path, root_dir)

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
            lines = content.split('\n')
    except Exception:
        return violations

    # Pattern 1: Icon buttons without aria-label
    icon_button_pattern = r'<(Button|IconButton|button)[^>]*(?:size\s*=\s*["\']icon["\']|className[^>]*icon)[^>]*>'
    for match in re.finditer(icon_button_pattern, content, re.IGNORECASE):
        line_num = content[:match.start()].count('\n') + 1
        element_content = match.group(0)

        # Check if aria-label exists in the same element or nearby
        if not re.search(r'aria-label\s*=', element_content, re.IGNORECASE):
            # Check next few lines
            start_line = max(0, line_num - 1)
            end_line = min(len(lines), line_num + 5)
            nearby_content = '\n'.join(lines[start_line:end_line])

            if not re.search(r'aria-label\s*=', nearby_content, re.IGNORECASE):
                violations.append(Violation(
                    file=relative_path,
                    line=line_num,
                    element='IconButton/Button',
                    issue='Icon button missing aria-label',
                    severity='error',
                    recommendation='Add aria-label prop with descriptive text (e.g., "Close dialog", "Send message")',
                    code_snippet=element_content[:100]
                ))

    # Pattern 2: Input elements without labels
    input_pattern = r'<(input|textarea|select)([^>]*?)>'
    for match in re.finditer(input_pattern, content, re.IGNORECASE):
        line_num = content[:match.start()].count('\n') + 1
        element_type = match.group(1)
        attributes = match.group(2) or ''

        # Skip hidden inputs
        if re.search(r'type\s*=\s*["\']hidden["\']', attributes, re.IGNORECASE):
            continue

        # Check for aria-label, aria-labelledby, or id
        has_aria_label = re.search(r'aria-label\s*=', attributes, re.IGNORECASE)
        has_aria_labelled_by = re.search(r'aria-labelledby\s*=', attributes, re.IGNORECASE)
        has_id = re.search(r'id\s*=\s*["\']([^"\']+)["\']', attributes, re.IGNORECASE)

        # Check if there's a label element with htmlFor
        has_label = False
        if has_id:
            id_match = re.search(r'id\s*=\s*["\']([^"\']+)["\']', attributes, re.IGNORECASE)
            if id_match:
                input_id = id_match.group(1)
                label_pattern = rf'htmlFor\s*=\s*["\']{re.escape(input_id)}["\']'
                has_label = re.search(label_pattern, content, re.IGNORECASE) is not None

        if not has_aria_label and not has_aria_labelled_by and not has_label:
            violations.append(Violation(
                file=relative_path,
                line=line_num,
                element=element_type,
                issue='Form input missing label or aria-label',
                severity='error',
                recommendation='Add htmlFor/id relationship, aria-label, or aria-labelledby',
                code_snippet=match.group(0)[:100]
            ))

    # Pattern 3: Modal/Dialog elements without proper ARIA
    modal_pattern = r'<(Dialog|DialogContent|Modal|dialog)([^>]*?)>'
    for match in re.finditer(modal_pattern, content, re.IGNORECASE):
        line_num = content[:match.start()].count('\n') + 1
        element_type = match.group(1)
        attributes = match.group(2) or ''

        issues = []
        if not re.search(r'role\s*=\s*["\']dialog["\']', attributes, re.IGNORECASE) and \
           not re.search(r'role\s*=\s*["\']alertdialog["\']', attributes, re.IGNORECASE):
            # Check if it's using Radix UI Dialog (which handles ARIA automatically)
            if 'DialogPrimitive' not in content[:match.start()]:
                issues.append('Missing role="dialog"')

        if not re.search(r'aria-modal\s*=\s*["\']true["\']', attributes, re.IGNORECASE):
            if 'DialogPrimitive' not in content[:match.start()]:
                issues.append('Missing aria-modal="true"')

        if not re.search(r'aria-labelledby\s*=', attributes, re.IGNORECASE):
            if 'DialogPrimitive' not in content[:match.start()]:
                issues.append('Missing aria-labelledby')

        if issues:
            violations.append(Violation(
                file=relative_path,
                line=line_num,
                element=element_type,
                issue=f'Modal/dialog missing ARIA attributes: {", ".join(issues)}',
                severity='error',
                recommendation='Add role="dialog", aria-modal="true", aria-labelledby, and optionally aria-describedby',
                code_snippet=match.group(0)[:150]
            ))

    # Pattern 4: Links without accessible text
    link_pattern = r'<(a|Link)([^>]*?)>(.*?)(?:</a>|</Link>)'
    for match in re.finditer(link_pattern, content, re.IGNORECASE | re.DOTALL):
        line_num = content[:match.start()].count('\n') + 1
        attributes = match.group(2) or ''
        children = match.group(3) or ''

        # Skip if it has aria-label
        if re.search(r'aria-label\s*=', attributes, re.IGNORECASE):
            continue

        # Check if link has visible text or image with alt text
        text_content = re.sub(r'<[^>]+>', '', children).strip()
        has_text = len(text_content) >= 2 and re.search(r'[A-Za-z0-9]{2,}', text_content)
        has_image_with_alt = re.search(r'<img[^>]*alt\s*=\s*["\'][^"\']+["\']', children, re.IGNORECASE)

        if not has_text and not has_image_with_alt:
            violations.append(Violation(
                file=relative_path,
                line=line_num,
                element='Link',
                issue='Link missing aria-label or accessible text',
                severity='warning',
                recommendation='Add aria-label prop or ensure link has visible text or image with alt text',
                code_snippet=match.group(0)[:150]
            ))

    return violations

def generate_report(results: Dict[str, Any]) -> str:
    """Generate markdown report."""
    total_components = results['total_components']
    violations = results['violations']
    warnings = results['warnings']
    by_component = results['by_component']

    error_count = len(violations)
    warning_count = len(warnings)

    compliance_rate = ((total_components - error_count) / total_components * 100) if total_components > 0 else 0

    report = f"""# Web ARIA Labels Audit Report

**Generated**: {results['timestamp']}

## Summary

- **Total Components Audited**: {total_components}
- **Total Violations**: {error_count} errors, {warning_count} warnings
- **Compliance Rate**: {compliance_rate:.1f}%

"""

    if error_count == 0 and warning_count == 0:
        report += "## ✅ All Clear\n\nAll components have proper ARIA labels and accessibility attributes.\n\n"
        return report

    if error_count > 0:
        report += f"## Violations by Severity\n\n### Errors ({error_count})\n\n"
        report += "| File | Line | Element | Issue | Recommendation |\n"
        report += "|------|------|---------|-------|----------------|\n"

        for violation in violations:
            file_display = violation['file'].replace('apps/web/src/components/', '')
            report += f"| `{file_display}` | {violation['line']} | `{violation['element']}` | {violation['issue']} | {violation['recommendation']} |\n"
        report += "\n"

    if warning_count > 0:
        report += f"### Warnings ({warning_count})\n\n"
        report += "| File | Line | Element | Issue | Recommendation |\n"
        report += "|------|------|---------|-------|----------------|\n"

        for warning in warnings:
            file_display = warning['file'].replace('apps/web/src/components/', '')
            report += f"| `{file_display}` | {warning['line']} | `{warning['element']}` | {warning['issue']} | {warning['recommendation']} |\n"
        report += "\n"

    report += """## Recommendations

1. **Icon Buttons**: All icon-only buttons must have `aria-label` prop
2. **Form Inputs**: All inputs must have associated labels via `htmlFor`/`id` or `aria-label`
3. **Modals/Dialogs**: Must have `role="dialog"`, `aria-modal="true"`, and `aria-labelledby`
4. **Interactive Elements**: Buttons, links, and custom interactive elements need `aria-label` or visible text
5. **Live Regions**: Dynamic content should use `aria-live` regions for screen reader announcements

"""

    return report

def main():
    """Main function."""
    print("🔍 Starting Web ARIA Labels Audit...\n")

    root_dir = os.path.join(os.getcwd(), 'apps/web/src/components')
    files = find_component_files(root_dir)
    print(f"Found {len(files)} component files to audit\n")

    all_violations = []
    by_component = {}

    for file_path in files:
        violations = audit_file(file_path, os.path.join(os.getcwd(), 'apps/web/src/components'))

        if violations:
            component_name = os.path.basename(file_path).replace('.tsx', '')
            by_component[component_name] = {
                'violations': [asdict(v) for v in violations if v.severity == 'error'],
                'warnings': [asdict(v) for v in violations if v.severity == 'warning'],
            }
            all_violations.extend(violations)

    errors = [asdict(v) for v in all_violations if v.severity == 'error']
    warnings = [asdict(v) for v in all_violations if v.severity == 'warning']

    results = {
        'timestamp': str(Path(__file__).stat().st_mtime),
        'total_components': len(files),
        'violations': errors,
        'warnings': warnings,
        'by_component': by_component,
    }

    print(f"✅ Audit complete!\n")
    print(f"- Total violations: {len(errors)} errors, {len(warnings)} warnings\n")

    # Generate report
    report = generate_report(results)
    report_dir = os.path.join(os.getcwd(), 'docs/accessibility')
    os.makedirs(report_dir, exist_ok=True)
    report_path = os.path.join(report_dir, 'web-aria-audit-report.md')

    with open(report_path, 'w', encoding='utf-8') as f:
        f.write(report)

    print(f"📄 Report saved to: {report_path}\n")

    if errors:
        print(f"❌ Found {len(errors)} errors that need to be fixed.\n")
    if warnings:
        print(f"⚠️  Found {len(warnings)} warnings to review.\n")
    if not errors and not warnings:
        print("✅ No violations found! All components are compliant.\n")

if __name__ == '__main__':
    main()
